File: 695-max-area-of-island/695-max-area-of-island.ts
function maxAreaOfIsland(grid: number[][]): number {
    let ans = 0;
    let dimensions = [grid.length, grid[0].length];
    
    for(let i = 0; i < dimensions[0]; i++) {
        for(let j = 0; j < dimensions[1]; j++) {
            if(grid[i][j] == 1){
               let res = dfs(i, j, grid, dimensions[0], dimensions[1]);
                ans = Math.max(ans, res);
            }
        }
    }
    return ans;
};

function dfs(row :number, col:number, grid: number[][], m: number, n : number) {
    if(row < 0 || row >= m || col < 0 || col >= n || grid[row][col] === 0) {
        return 0;
    }
    
    grid[row][col] = 0;
    
    return 1 + dfs(row+1, col, grid, m, n) + dfs(row-1, col, grid, m, n) + dfs(row, col+1, grid, m, n) + dfs(row, col-1, grid, m, n);
}

File: 695-max-area-of-island/695-max-area-of-island.java
class Solution {
    public int islandArea = 0;
    public int maxAreaOfIsland(int[][] grid) {
        int M = grid.length;
        int N = grid[0].length;
        boolean[][] visited = new boolean[M][N];
        int maxArea = 0;
        for(int i = 0; i < M; i++){
            for(int j = 0; j < N; j++){
                islandArea = 0;
                if(!visited[i][j] && grid[i][j] == 1){
                    dfs(i,j, visited, grid, 0);
                    maxArea = Math.max(maxArea, islandArea);
                }
                visited[i][j] = true;
            }
        }
        return maxArea; 
    }
    public void dfs(int row, int col, boolean[][]visited,int[][] grid, int size){
        int M = grid.length;
        int N = grid[0].length;
        if(row >= M || row < 0 || col >= N || col < 0 || visited[row][col] || grid[row][col] == 0){
            return;
        }
        visited[row][col] = true;
        islandArea++;
        dfs(row-1,col, visited, grid, size); 
        dfs(row+1,col, visited, grid, size);
        dfs(row,col-1, visited, grid, size);
        dfs(row,col+1, visited, grid, size);
    }
}